framework/src/transfers/MultiAppGeneralFieldTransfer.C

```diff
@@ -0,0 +1,831 @@
+  }
+}
+
+void
+MultiAppGeneralFieldTransfer::buildMeshFunctions(
+    const VariableName & var_name, std::vector<std::shared_ptr<MeshFunction>> & local_meshfuns)
+{
+  // Construct a local mesh for each problem
+  for (unsigned int i_from = 0; i_from < _from_problems.size(); ++i_from)
+  {
+    FEProblemBase & from_problem = *_from_problems[i_from];
+    MooseVariableFEBase & from_var = from_problem.getVariable(
+        0, var_name, Moose::VarKindType::VAR_ANY, Moose::VarFieldType::VAR_FIELD_STANDARD);
+
+    System & from_sys = from_var.sys().system();
+    unsigned int from_var_num = from_sys.variable_number(from_var.name());
+
+    std::shared_ptr<MeshFunction> from_func;
+    from_func.reset(new MeshFunction(
+        from_problem.es(), *from_sys.current_local_solution, from_sys.get_dof_map(), from_var_num));
+    from_func->init();
+    from_func->enable_out_of_mesh_mode(BetterOutOfMeshValue);
+    local_meshfuns.push_back(from_func);
+  }
+}
+
+void
+MultiAppGeneralFieldTransfer::evaluateInterpValues(
+    const std::vector<BoundingBox> & local_bboxes,
+    const std::vector<std::shared_ptr<MeshFunction>> & local_meshfuns,
+    const std::vector<Point> & incoming_points,
+    std::vector<Real> & outgoing_vals)
+{
+  dof_id_type i_pt = 0;
+  for (auto & pt : incoming_points)
+  {
+    // Loop until we've found the lowest-ranked app that actually contains
+    // the quadrature point.
+    for (MooseIndex(_from_problems.size()) i_from = 0;
+         i_from < _from_problems.size() && outgoing_vals[i_pt] == BetterOutOfMeshValue;
+         ++i_from)
+    {
+      if (local_bboxes[i_from].contains_point(pt))
+      {
+        // Use mesh funciton to compute interpolation values
+        auto val = (*local_meshfuns[i_from])(pt - _from_positions[i_from]);
+        // Assign value
+        outgoing_vals[i_pt] = val;
+      }
+    }
+
+    // Move to next point
+    i_pt++;
+  }
+}
+
+void
+MultiAppGeneralFieldTransfer::cacheIncomingInterpVals(
+    processor_id_type pid,
+    const VariableName & var_name,
+    std::vector<PointInfo> & pointInfoVec,
+    const std::vector<Point> & point_requests,
+    const std::vector<Real> & incoming_vals,
+    DofobjectToInterpValVec & dofobject_to_valsvec,
+    InterpCaches & interp_caches)
+{
+  mooseAssert(pointInfoVec.size() == incoming_vals.size(),
+              " Number of dof objects does not equal to the number of incoming values");
+
+  dof_id_type val_offset = 0;
+  for (auto & pointinfo : pointInfoVec)
+  {
+    const auto problem_id = pointinfo.problem_id;
+    const auto dof_object_id = pointinfo.dof_object_id;
+
+    const std::pair<unsigned int, dof_id_type> dofobject(problem_id, dof_object_id);
+
+    // libMesh EquationSystems
+    auto & es = _to_problems[problem_id]->es();
+    // libMesh system
+    System * to_sys = find_sys(es, var_name);
+
+    // libMesh mesh
+    MeshBase & to_mesh = _to_meshes[problem_id]->getMesh();
+    auto var_num = to_sys->variable_number(var_name);
```

Comment:
Need to not use strings for indexing

Reply:
@friedmud so store the variable numbers directly then?
the variable names are used in all the routines currently, you d rather the var num?